pkg/agent/core/metrics.go

```diff
@@ -0,0 +1,182 @@
+	// DeepCopy produces a copy of the value, intended for use when serializing State
+	DeepCopy() ScalingAlgorithm[M]
+}
+
+// ToApiMetrics is the interface that ScalingAlgorithm's metrics must satisfy - namely, that they
+// can be converted to an api.Metrics
+type ToAPIMetrics interface {
+	ToAPI() api.Metrics
+}
+
+// SimpleFactorScaling is a ScalingAlgorithm implementation that calculates the desired resource
+// usage by a simple multiple of the most recently collected metrics
+type SimpleFactorScaling struct {
+	// nb: fields only public to allow JSON serialization
+	Metrics *Metrics
+}
+
+func NewSimpleFactorScaling() *SimpleFactorScaling {
+	return &SimpleFactorScaling{
+		Metrics: nil,
+	}
+}
+
+// DeepCopy implements ScalingAlgorithm[Metrics]
+func (s *SimpleFactorScaling) DeepCopy() ScalingAlgorithm[Metrics] {
+	return &SimpleFactorScaling{
+		Metrics: shallowCopy[Metrics](s.Metrics),
+	}
+}
+
+// Update implements ScalingAlgorithm[Metrics]
+func (s *SimpleFactorScaling) Update(metrics Metrics) {
+	s.Metrics = &metrics
+}
+
+// GoalCU implements ScalingAlgorithm[Metrics], refer there for more.
+func (s *SimpleFactorScaling) GoalCU(scalingConfig api.ScalingConfig, computeUnit api.Resources) (_ uint32, ok bool) {
+	if s.Metrics == nil {
+		return 0, false
+	}
+	// Broadly, the implementation works like this:
+	// For CPU:
+	// Based on load average, calculate the "goal" number of CPUs (and therefore compute units)
+	//
+	// For Memory:
+	// Based on memory usage, calculate the VM's desired memory allocation and extrapolate a
+	// goal number of CUs from that.
+	//
+	// 1. Take the maximum of these two goal CUs to create a unified goal CU
+	// 2. Cap the goal CU by min/max, etc
+	// 3. that's it!
+
+	// For CPU:
+	// Goal compute unit is at the point where (CPUs) × (LoadAverageFractionTarget) == (load
+	// average),
+	// which we can get by dividing LA by LAFT, and then dividing by the number of CPUs per CU
+	goalCPUs := float64(s.Metrics.LoadAverage1Min) / scalingConfig.LoadAverageFractionTarget
+	cpuGoalCU := uint32(math.Round(goalCPUs / computeUnit.VCPU.AsFloat64()))
+
+	// For Mem:
+	// Goal compute unit is at the point where (Mem) * (MemoryUsageFractionTarget) == (Mem Usage)
+	// We can get the desired memory allocation in bytes by dividing MU by MUFT, and then convert
+	// that to CUs
+	//
+	// NOTE: use uint64 for calculations on bytes as uint32 can overflow
+	memGoalBytes := uint64(math.Round(float64(s.Metrics.MemoryUsageBytes) / scalingConfig.MemoryUsageFractionTarget))
+	memGoalCU := uint32(memGoalBytes / uint64(computeUnit.Mem))
+
+	goalCU := util.Max(cpuGoalCU, memGoalCU)
+	return goalCU, true
+}
+
+// ReadMetrics generates Metrics from node_exporter output, or returns error on failure
+//
+// This function could be more efficient, but realistically it doesn't matter. The size of the
+// output from node_exporter/vector is so small anyways.
+func ReadMetrics(nodeExporterOutput []byte, loadPrefix string) (m Metrics, err error) {
```

Comment:
Can you add an example input to know what does it parse?

Feels like this parsing should belong somewhere else. Is this standard prometheus format? Will
https://pkg.go.dev/github.com/prometheus/common/expfmt work?

Reply:
I'll add some tests.

For context, this was directly copied from `main`'s current `pkg/api/metrics.go`.

Yeah, this is standard prometheus format. I _think_ I took a look at that package while prepping this PR, but it seemed complicated and better to leave as a follow-up :sweat_smile: